File: go/logger.go
package realmid

import (
	"io"
	"log/slog"
)

// noopLogger returns a slog logger that discards all output. Used as the
// default when the partner does not supply one (SPEC §9).
func noopLogger() *slog.Logger {
	return slog.New(slog.NewTextHandler(io.Discard, nil))
}

// redactCredential truncates a bearer-shaped string to its first 6
// characters, suffixed with `…`. Returns "<empty>" for empty input so
// log readers can still distinguish "no credential" from "redacted".
func redactCredential(value string) string {
	if value == "" {
		return "<empty>"
	}
	if len(value) <= 6 {
		return value + "…"
	}
	return value[:6] + "…"
}

File: ts/src/logger.ts
/**
 * Structured logger interface — SPEC §9.
 *
 * A minimal, no-dependency contract that `console`, `pino`, `bunyan`, and
 * any other modern Node logger satisfies out of the box. The SDK never
 * logs raw credentials; helpers in this module redact bearer-shaped
 * strings to their first 6 characters.
 */

export interface Logger {
  debug(msg: string, meta?: Record<string, unknown>): void;
  info(msg: string, meta?: Record<string, unknown>): void;
  warn(msg: string, meta?: Record<string, unknown>): void;
  error(msg: string, meta?: Record<string, unknown>): void;
}

/** Default no-op logger used when the partner does not supply one. */
export const NOOP_LOGGER: Logger = {
  debug() {},
  info() {},
  warn() {},
  error() {},
};

/**
 * Truncate a credential to the first 6 chars, suffixed with `…`. Returns
 * the literal string `"<empty>"` for empty input so callers can still
 * tell apart "no credential" from "credential redacted".
 */
export function redactCredential(value: string | undefined | null): string {
  if (!value) return "<empty>";
  if (value.length <= 6) return value + "…";
  return value.slice(0, 6) + "…";
}
